precompile/contract_native_minter.go

```diff
@@ -0,0 +1,133 @@
+}
+
+// GetContractNativeMinterStatus returns the role of [address] for the minter list.
+func GetContractNativeMinterStatus(stateDB StateDB, address common.Address) AllowListRole {
+	return getAllowListStatus(stateDB, ContractNativeMinterAddress, address)
+}
+
+// SetContractNativeMinterStatus sets the permissions of [address] to [role] for the
+// contract deployer allow list.
+// assumes [role] has already been verified as valid.
+func SetContractNativeMinterStatus(stateDB StateDB, address common.Address, role AllowListRole) {
+	setAllowListRole(stateDB, ContractNativeMinterAddress, address, role)
+}
+
+// PackMintInput packs [address] and [amount] into the appropriate arguments for minting operation.
+func PackMintInput(address common.Address, amount *big.Int) ([]byte, error) {
+	// function selector (4 bytes) + input(hash for address + hash for amount)
+	input := make([]byte, 0, selectorLen+inputLen)
+	input = append(input, mintSignature...)
+	input = append(input, address.Hash().Bytes()...)
+	input = append(input, amount.Bytes()...)
+	return input, nil
+}
+
+// UnpackMintInput attempts to unpack [input] into the arguments to the mint precompile
+func UnpackMintInput(input []byte) (common.Address, *big.Int, error) {
+	if len(input) != inputLen {
+		return common.Address{}, nil, fmt.Errorf("invalid input length for minting: %d", len(input))
+	}
+	to := common.BytesToAddress(input[:common.AddressLength])
+	assetAmount := new(big.Int).SetBytes(input[common.AddressLength : common.AddressLength+common.HashLength])
+	return to, assetAmount, nil
+}
+
+// createMint // TODO: add comment
+func createMint(precompileAddr common.Address) RunStatefulPrecompileFunc {
+	return func(evm PrecompileAccessibleState, callerAddr, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+		// Note: this should never happen since the required gas should be verified before calling Run.
+		if suppliedGas < MintGasCost {
+			return nil, 0, fmt.Errorf("%w (%d) < (%d)", ErrExceedsGasAllowance, MintGasCost, suppliedGas)
+		}
+		remainingGas = suppliedGas - MintGasCost
+
+		if !evm.GetChainRules().ContractNativeMinterEnabled() {
+			return nil, remainingGas, fmt.Errorf("native minter is not activated")
+		}
+
+		if readOnly {
+			return nil, remainingGas, ErrWriteProtection
+		}
+
+		to, amount, err := UnpackMintInput(input)
+		if err != nil {
+			return nil, remainingGas, err
+		}
+
+		// Verify that the caller is in the allow list and therefore has the right to modify it
+		callerStatus := getAllowListStatus(evm.GetStateDB(), precompileAddr, callerAddr)
+		if !callerStatus.IsEnabled() {
+			return nil, remainingGas, fmt.Errorf("%w: %s", ErrCannotMint, callerAddr)
+		}
+
+		if !evm.GetStateDB().Exist(to) {
+			if remainingGas < CallNewAccountGas {
+				return nil, 0, fmt.Errorf("%w (%d) < (%d)", ErrExceedsGasAllowance, CallNewAccountGas, suppliedGas)
+			}
+			remainingGas -= CallNewAccountGas
```

Comment:
This makes our `RequiredGas` invariant weird because it could surpass the amount returned?

Reply:
ah, true! however the `RequiredGas` function of this precompile requires evm state (for `state.Exist(addrs)`) should we change `RequiredGas` structure to take state. I think we need to re-implement what we already have in `createMint` to determine required gas. Alternatively we can omit `RequiredGas` function altogether and let gas checks run in the execute functions. @aaronbuchwald

Reply:
I'd be in favor of adding state into `RequiredGas`. The reason I separated them was so that we could return `ErrOutOfGas` if there was not sufficient gas to cover the operation.

Alternatively, I think the best solution would be to move all of the errors into a package that could be used by both the EVM and the precompile package, so that we can return `ErrOutOfGas` from the precompile directly, and then I think it would make sense to turn the stateful precompiles into a single execution function and get rid of the `RequiredGas` function entirely.

Reply:
removed `RequiredGas`. We can return  `ErrOutOfGas` without moving it into a separate package. Returned it in the helper function.